### PR TITLE
Fixed frame converter

### PR DESCRIPTION
**Pull request by kaktus40**
In continuity of https://github.com/AnalyticalGraphicsInc/cesium/pull/4498 and in response for https://github.com/AnalyticalGraphicsInc/cesium/issues/3256 , I implemented an ability to express a rotation around a local frame different of ENU.

### CODE REVIEW

**Comment by pjcozzi** [comment]
Thanks for this big contribution, @kaktus40!

@bagnell could you please review this?

@xtassin @speigg your input here is also very much appreciated.

**Comment by pjcozzi** [comment]
@kaktus40 could you also please update CHANGES.md and resize the new images in the `Apps/Sandcastle/gallery/` directory to be the same as the other Sandcastle thumbnails?

**Comment by kaktus40** [comment]
@pjcozzi :it's done

**Comment by kaktus40** [comment]
Hello, after playing with LocalToFixedFrame gallery example, I fear there is another misconception in Cesium. Indeed, heading value is the rotation angle around the third axis (for example in East North Up local reference frame, it's the rotation angle around "Up").   As the reference frame is direct, a positive value of heading means a rotation from first axis (respectively east) to second axis (respectively north) in the plane generated by first and second axis, see [here](https://en.wikipedia.org/wiki/File:MISB_ST_0601.8_-_Yaw,_Pitch_%26_Roll.png). In LocalToFixedFrame gallery example when you turn heading positively, we should expect that the model turns from east to north in ENU representation but it's the opposite. After checking the code I found [this](https://github.com/AnalyticalGraphicsInc/cesium/blob/master/Source/Core/Quaternion.js#L197): I don't understand why it's -heading that is used and not simply heading. It's the same remark for pitch [here](https://github.com/AnalyticalGraphicsInc/cesium/blob/master/Source/Core/Quaternion.js#L195)

**Comment by kaktus40** [comment]
Hello, any news for this PR?

**Comment by hpinkos** [comment]
Sorry we haven't had a chance to look at this yet @kaktus40.  We might have a chance to at the bug bash this week

File: Source/Core/Transforms.js [hunk 1]
```diff
@@ -51,111 +53,187 @@ define([
...
+        west: new Cartesian3(),
+        south: new Cartesian3(),
+        down: new Cartesian3()
+      };
+      var firstCartesian = new Cartesian3();
+      var secondCartesian = new Cartesian3();
+      var thirdCartesian = new Cartesian3();
+
+      /**
+       * [resultat description]
```
Comment (bagnell) [inline Source/Core/Transforms.js]: You can create function doc for the returned function. See [here](https://github.com/AnalyticalGraphicsInc/cesium/blob/master/Source/Scene/Billboard.js#L1290) for an example.

Comment (bagnell) [inline Source/Core/Transforms.js]: @mramato Do we still use `@callback` for this? Is there a way to generate doc if we generate all combinations up front?

File: Source/Core/Transforms.js [hunk 2]
```diff
@@ -473,16 +328,20 @@ define([
      * var hpr = new Cesium.HeadingPitchRoll(heading, pitch, roll);
      * var transform = Cesium.Transforms.headingPitchRollToFixedFrame(center, hpr);
      */
-    Transforms.headingPitchRollToFixedFrame = function(origin, headingPitchRoll, ellipsoid, result) {
-        Check.typeOf.object('headingPitchRoll', headingPitchRoll);
-        var heading = headingPitchRoll.heading;
-        var pitch = headingPitchRoll.pitch;
-        var roll = headingPitchRoll.roll;
+    Transforms.headingPitchRollToFixedFrame = function(origin, headingPitchRoll, ellipsoid, fixedFrameTransformOrResult, result) {
+        Check.typeOf.object( 'HeadingPitchRoll', headingPitchRoll);
```
Comment (bagnell) [inline Source/Core/Transforms.js]: This should go inside the pragmas to be removed from release builds.

File: Source/Core/Transforms.js [hunk 3]
```diff
@@ -473,16 +328,20 @@ define([
...
 
         // checks for required parameters happen in the called functions
-        var hprQuaternion = Quaternion.fromHeadingPitchRoll(heading, pitch, roll, scratchHPRQuaternion);
+        deprecationWarning('Transforms.headingPitchRollToFixedFrame(origin, headingPitchRoll, ellipsoid, result)', 'The method was deprecated in Cesium 1.31 and will be removed in version 1.32. ' +
```
Comment (bagnell) [inline Source/Core/Transforms.js]: Same comment.

File: Source/Core/Transforms.js [hunk 4]
```diff
@@ -510,10 +371,16 @@ define([
      * var hpr = new HeadingPitchRoll(heading, pitch, roll);
      * var quaternion = Cesium.Transforms.headingPitchRollQuaternion(center, hpr);
      */
-    Transforms.headingPitchRollQuaternion = function(origin, headingPitchRoll, ellipsoid, result) {
+    Transforms.headingPitchRollQuaternion = function(origin, headingPitchRoll, ellipsoid, fixedFrameTransformOrResult, result) {
         // checks for required parameters happen in the called functions
-        Check.typeOf.object('headingPitchRoll', headingPitchRoll);
-        var transform = Transforms.headingPitchRollToFixedFrame(origin, headingPitchRoll, ellipsoid, scratchENUMatrix4);
+        Check.typeOf.object( 'HeadingPitchRoll', headingPitchRoll);
```
Comment (bagnell) [inline Source/Core/Transforms.js]: Same comments.

File: Source/Scene/Camera.js [hunk 5]
```diff
@@ -1160,14 +1164,16 @@ define([
         var pitch = defaultValue(orientation.pitch, -CesiumMath.PI_OVER_TWO);
         var roll = defaultValue(orientation.roll, 0.0);
 
+        var hpr = new HeadingPitchRoll(heading, pitch, roll);
```
Comment (bagnell) [inline Source/Scene/Camera.js]: Create a scratch variable.

File: Specs/Core/QuaternionSpec.js [hunk 6]
```diff
@@ -652,7 +660,7 @@ defineSuite([
             Quaternion.fromRotationMatrix(undefined);
         }).toThrowDeveloperError();
     });
-
+/*
```
Comment (bagnell) [inline Specs/Core/QuaternionSpec.js]: Did you mean to leave these commented out?

**Comment by pjcozzi** [comment]
> Does anyone else want to review this?

Just @bagnell's review is OK with me.

@kaktus40 if you are able to update this tomorrow or Friday morning, it will likely get merged quickly since we are having a bug bash and focused on merging PRs.

Thanks again for this!

File: Source/Core/Quaternion.js [hunk 7]
```diff
@@ -184,17 +188,29 @@ define([
      * @param {Quaternion} [result] The object onto which to store the result.
      * @returns {Quaternion} The modified result parameter or a new Quaternion instance if none was provided.
      */
-    Quaternion.fromHeadingPitchRoll = function(heading, pitch, roll, result) {
+    Quaternion.fromHeadingPitchRoll = function(headingOrHeadingPitchRoll, pitchOrResult, roll, result) {
```
Comment (kaktus40) [inline Source/Core/Quaternion.js]: I don't see how to use options parameter in Quaternion.fromHeadingPitchRoll as the first parameter will be a HeadingPitchRoll object?
You suggest that rollQuaternion, pitchQuaternion and headingQuaternion become scratch variables? Ok for me but just for information that was the previous design.

File: Source/Core/Transforms.js [hunk 8]
```diff
@@ -51,111 +53,187 @@ define([
...
+      var thirdCartesian = new Cartesian3();
+
+      /**
+       * [resultat description]
+       * @param {Cartesian3} origin The center point of the local reference frame.
+       * @param {Ellipsoid} [ellipsoid=Ellipsoid.WGS84] The ellipsoid whose fixed frame is used in the transformation.
+       * @param {Matrix4} [result] The object onto which to store the result.
+       * @returns {Matrix4} The modified result parameter or a new Matrix4 instance if none was provided.
+       */
+      var resultat= function(origin, ellipsoid, result) {
```
Comment (kaktus40) [inline Source/Core/Transforms.js]: Sorry I don't see what you mean :-(

File: Specs/Core/QuaternionSpec.js (hunk 6 above)
Comment (kaktus40) [inline Specs/Core/QuaternionSpec.js]: As fromHeadingPitchRoll use now a headingPitchRoll class, the commented tests have no interest.

**Comment by kaktus40** [comment]
Ok I corrected the code in compliance with the review code

File: Source/Core/Transforms.js (hunk 8 above)
Comment (bagnell) [inline Source/Core/Transforms.js]: This function will be generated each time `localFrameToFixedFrameGenerator` is called. So every call like: `Cesium.Transforms.localFrameToFixedFrameGenerator('north','west')` will generate a new function but it doesn't change. Create a file scope object like `localFrameToFixedFrameCache` with unique names based on the axis. First you check if the function is in the cache, if it is, return it; if not, generate it and place it in the cache before returning it.

File: Source/Core/Quaternion.js [hunk 9]
```diff
@@ -184,18 +191,30 @@ define([
...
         //>>includeStart('debug', pragmas.debug);
-        Check.typeOf.number('heading', heading);
-        Check.typeOf.number('pitch', pitch);
-        Check.typeOf.number('roll', roll);
+        if (headingOrHeadingPitchRoll instanceof HeadingPitchRoll) {
+          Check.typeOf.object('headingPitchRoll',headingOrHeadingPitchRoll );
```
Comment (bagnell) [inline Source/Core/Quaternion.js]: There are still whitespace issues. Four spaces should be used for each indent. Please update throughout.

**Comment by kaktus40** [comment]
I made the corrections

**Comment by bagnell** [comment]
Thanks @kaktus40!

**Pull request merged**